### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by ankitrex**

### CODE REVIEW

File: src/main/java/com/qwerty/ankit/ArrayCalculations.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+		Arrays.sort(arr);
+		int arrayLength = arr.length;
+
+		for (int i = 0; i < arrayLength - 2; i++) {
+			int j = i + 1;
+			int k = arrayLength - 1;
+			while (j != k) {
+				int total = arr[i] + arr[j] + arr[k];
+				if (total == sum)
+					return arr[i] + "," + arr[j] + "," + arr[k];
```
Comment (ankuriitk) [inline src/main/java/com/qwerty/ankit/ArrayCalculations.java]: the formatting of output should be separated out into a separate method. Ideally the combination finding method should just return array. That way if output format needs to be changed, only the formatting method needs to be changed and not the logic of finding subset

File: src/main/java/com/qwerty/ankit/Friend.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package com.qwerty.ankit;
+
+public class Friend {
+
+	private String name;
+	private String username;
+	private double distance;
+	
+	public Friend(String name, String username, double distance) {
+		super();
```
Comment (ankuriitk) [inline src/main/java/com/qwerty/ankit/Friend.java]: appreciate the use of classes! However, the code would have been more cleaner if you had moved distance calculation here or perhaps define a LatLong class. Most of the main code is in PartyInvitations, the abstraction and modularity benefit of classes is lost in the current design.

File: src/main/java/com/qwerty/ankit/PartyInvitations.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class PartyInvitations {
+
+	private static final int earthRadius = 6371;
+	private static final double myLatitude = 12.9611159;
+	private static final double myLongitude = 77.6362214;
```
Comment (ankuriitk) [inline src/main/java/com/qwerty/ankit/PartyInvitations.java]: excellent - many candidates embed constants in the code logic itself. It is always a good idea to define them at top of file and use in code.

One nit here: while the question was certainly about Clarisights office location, the code be easily generalized to take any home location. As such, rather than fixing them, parameterizing them and letting main() pass them as args would have been even better

File: src/main/java/com/qwerty/ankit/PartyInvitations.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+		List<Friend> friends = new ArrayList<>();
+
+		for (int i = 0; i < allFriends.length(); i++) {
+			JSONObject friend;
+			try {
+				friend = allFriends.getJSONObject(i);
+				double distance = distanceInKms(myLatitude, myLongitude, friend.optDouble("latitude"),
+						friend.optDouble("longitude"));
+				if (distance <= 100) {
```
Comment (ankuriitk) [inline src/main/java/com/qwerty/ankit/PartyInvitations.java]: ideally this 100 should also have been defined as a constant at the top

File: src/main/java/com/qwerty/ankit/PartyInvitations.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+				sb.append(line);
+			}
+			return sb.toString();
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+		return null;
+	}
+
+	private double distanceInKms(double startLatitude, double startLongitude, double endLatitude,
```
Comment (ankuriitk) [inline src/main/java/com/qwerty/ankit/PartyInvitations.java]: best location for this method would have been a LatLng class

File: src/main/java/com/qwerty/ankit/PartyInvitations.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+					friends.add(new Friend(friend.optString("name"), friend.optString("username"), distance));
+				}
+			} catch (JSONException e) {
+				e.printStackTrace();
+			}
+		}
+		return friends;
+	}
+
+	private String readFileAsString(String path) {
```
Comment (ankuriitk) [inline src/main/java/com/qwerty/ankit/PartyInvitations.java]: while this was not a requirement and no points will be deducted for this, it is a bad idea to read entire file in one shot. What if the file is 1GB in size? At Clarisights, we have to deal with really large files, reading entirely into memory will kill our jobs!

File: src/test/java/com/qwerty/ankit/ArrayCalculationsTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+	private ArrayCalculations arrayCalculations = new ArrayCalculations();
+
+	@Test
+	public void threeElemCombinationsTest1() {
+
+		int arr[] = { 1, 4, 5, 6, 2, 10 };
+		int sum = 13;
+		List<String> validCombinations = Arrays.asList("1,2,10", "2,5,6");
+		String combination = arrayCalculations.threeElemCombination(arr, sum);
```
Comment (ankuriitk) [inline src/test/java/com/qwerty/ankit/ArrayCalculationsTest.java]: good that you output all combinations

File: src/test/java/com/qwerty/ankit/ArrayCalculationsTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	@Test
+	public void threeElemCombinationsTest5() {
+
+		int arr[] = {};
+		int sum = 0;
+
+		String combination = arrayCalculations.threeElemCombination(arr, sum);
+
+		Assert.assertTrue("NA".equals(combination));
+	}
```
Comment (ankuriitk) [inline src/test/java/com/qwerty/ankit/ArrayCalculationsTest.java]: no test for negative numbers?

File: src/main/java/com/qwerty/ankit/PartyInvitations.java (hunk 4 above)
Comment (ankuriitk) [reply src/main/java/com/qwerty/ankit/PartyInvitations.java]: you have a bug here - your variable range is not used in the method at all. This 100 should have been replaced by range.

File: src/test/java/com/qwerty/ankit/PartyInvitationsTest.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+	
+	private static final int rangeInKms = 100;
+	
+	@Test
+	public void inviteMyFriendsTest1(){
+		
+		String filePath = "src/test/resources/friends1.json";
+		String expectedResult = "Kevin-f6 :: Stanley-f8 :: Michael-f3 :: Jim-f1 :: Pam-f2 :: Phyllis-f9 :: Oscar-f7 :: Angela-f5 :: Erin-f11 :: Ryan-f12 :: Kelly-f13";
+		
+		String result = partyInvitations.inviteMyFriends(filePath, rangeInKms);
```
Comment (ankuriitk) [inline src/test/java/com/qwerty/ankit/PartyInvitationsTest.java]: if you had tried with a value other than 100, you would have probably discovered the range variable bug in your code.

File: src/test/resources/friends1.json [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+		"latitude":12.1511559,
+		"longitude":75.9362214
+	},
+	{
+		"username":"f17",
+		"name":"Toby",
+		"latitude":11.1411559,
+		"longitude":77.3362214
+	}
+]
```
Comment (ankuriitk) [inline src/test/resources/friends1.json]: few edge cases missed: negative lat/lng, north pole/south pole checks

File: src/main/java/com/qwerty/ankit/ArrayCalculations.java (hunk 1 above)
Comment (ankitrex) [reply src/main/java/com/qwerty/ankit/ArrayCalculations.java]: Noted. Will keep that in mind.

File: src/main/java/com/qwerty/ankit/PartyInvitations.java (hunk 3 above)
Comment (ankitrex) [reply src/main/java/com/qwerty/ankit/PartyInvitations.java]: Agreed

File: src/main/java/com/qwerty/ankit/PartyInvitations.java (hunk 4 above)
Comment (ankitrex) [reply src/main/java/com/qwerty/ankit/PartyInvitations.java]: Ahh missed that in refactoring the code. Will fix it.

File: src/main/java/com/qwerty/ankit/PartyInvitations.java (hunk 6 above)
Comment (ankitrex) [reply src/main/java/com/qwerty/ankit/PartyInvitations.java]: Yea. Should've considered that.

**Comment by ankitrex** [comment]
Thanks for the feedback. Much appreciated.
